### PR TITLE
Fix #418: Prevent notifications from crashing the site on mobile and make audio work correctly across devices

### LINKED ISSUE #418: Problem when turning on notifications on Android Chrome
When I turn on notifications (either for a pokemon or rarity) and a pokemon is shown on the map that would trigger the notification the whole process stops there. No notification is shown, no other pokemons will be loaded on the map (but the ones that were loaded before the pokemon that raised the notification). It is like the site crashes or something. Removing the notification pokemon from the the notification list and reloading the site helps.

Using: Android Chrome Version 52.0.2743.98 on Android 5.1.1

On Desktop Chrome (Win) it is working

### PR DESCRIPTION
**Pull request by Studio-XP**
## Description
Fixes notifications from crashing the page on Android. I'm an ass, so I didn't realize that I shouldn't make pull requests on default branch on my forked version :/ Proper pull request for #461

## Motivation and Context
With notifications enabled while viewing on your phone, the page would crash, not showing any new Pokemon #338 #418. This commit prevents the page from crashing, while enabling mobile users to receive push notifications on their phone.

## How Has This Been Tested?
On both Chrome and Firefox for both desktop and Android.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by khag7** [comment]
Yay! I was just about to install Firefox on my phone to get mobile notifications going but I'll hold off for this to be merged into a release

**Comment by Studio-XP** [comment]
Only catch though is that you need to have this behind an ssl proxy, either Nginx or Apache, for the notifications to work on Chrome. Chrome on Android prevents notifications from an untrusted location (non https)

File: static/js/map.js [hunk 1]
```diff
@@ -1400,7 +1400,7 @@ function loadRawData () {
 
 function processPokemons (i, item) {
   if (!Store.get('showPokemon')) {
-    return false; // in case the checkbox was unchecked in the meantime.
+    return false // in case the checkbox was unchecked in the meantime.
```
Comment (Chlodochar) [inline static/js/map.js]: I get why this was done, but it is unrelated to change. `develop` now already has this fixed.

File: static/js/map.js [hunk 2]
```diff
@@ -1448,7 +1448,7 @@ function processPokestops (i, item) {
 
 function processGyms (i, item) {
   if (!Store.get('showGyms')) {
-    return false; // in case the checkbox was unchecked in the meantime.
+    return false // in case the checkbox was unchecked in the meantime.
```
Comment (Chlodochar) [inline static/js/map.js]: Same here.

File: static/js/sw.js [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (Chlodochar) [inline static/js/sw.js]: What is this doing?

File: pogom/app.py [hunk 4]
```diff
@@ -226,6 +227,9 @@ def get_stats(self):
                                valid_input=self.get_valid_stat_input()
                                )
 
+    def static_from_root(self):
+        return send_from_directory(self.static_folder + '/js', request.path[1:])
+
```
Comment (Chlodochar) [inline pogom/app.py]: Why is this needed?

**Comment by Studio-XP** [comment]
So in order for the service worker (mobile notifications) to work, there has to be a blank sw.js file located in the document root. The changes in the app.py loads that file in the document root

**Comment by Chlodochar** [comment]
@Studio-XP That sounds pretty awful, but if that's what needed to get this going I'm all :+1: with this. Could you just add the explanation you just gave to `sw.js` as a comment or before `self.route("/sw.js", methods=['GET'])(self.static_from_root)`. It seems to come out of nowhere at the moment.

With that knowledge, the code looks good :+1:.

**Comment by Chlodochar** [comment]
Tested on Android. Notifications crashed before and with this PR it doesn't anymore. Code makes sense now. :+1: from me. Another contributor must merge.

**Comment by khag7** [comment]
Works for me

**Comment by DiscoTim** [comment]
not only do they not crash, they actually work and the sound works.. :) good job!

**Comment by Studio-XP** [comment]
@DiscoTim Thanks for the help with the 2 libraries

**Comment by khag7** [comment]
I've been playing around with notifications and I've figured out what the deal is with the service worker file (sw.js)

We can use that to display notifications even when the browser is closed. This has pretty cool implications. On mobile, a user could have the browser completely closed but could still receive notifications. Is this something we should look into doing?

**Comment by Studio-XP** [comment]
I tried to look into how to open an address with it, but was unable to find out any info. Let me know what you find. I would love to be able to open the page with it like on desktop

**Comment by khag7** [comment]
my code is kind of a ridiculous mess and I don't have much more time to tweak it today.

So far I have it so that the notifications update every few seconds. This gives me a countdown on the timer and if my location changes the distanc updates as well. So as I walk around outside the nearby pokemon are in my notification bar (with icons!) and if i pull down it updates distance and time every few seconds.

When I implement the action buttons some other stuff breaks, but I'm working on figuring out what the deal is with that and how to fix it.

Once I get all that working, I want to move the code out of the map.js and into the sw.js so that it can run even if the browser is closed. This will probably be the most difficult step and that might be where I get lost. It definitely needs to be a separate PR but I just figured it's worth mentioning here. It involves the user getting an API key for google cloud messaging (GCM) (similar to getting a maps API key) and setting up runserver.py to send out periodic notifications (since the browser isn't open, map.js isn't loaded and isn't constantly refreshing data, the server has to send it to GCM)

**Comment by Chlodochar** [comment]
Sounds cool. Not for this pr but definitely worth looking into. A new issue for this would be appropriate.

**Comment by khag7** [comment]
@Chlodochar Yeah for sure, I'll wait til I have things a bit more "together" before I even open an issue. The only reason I brought it up here is that using a blank sw.js and putting notifications in the map.js is kind of hacky. The serviceworker js is there for a reason, we're skipping it entirely and even though it works we could definitely do better!

**Comment by DiscoTim** [comment]
its a very basic start to working notifications, no service worker means no notifications, or as it is now broken notifications, and a blank service worker is better than none. notifications could be extended forever, and i agree with @khag7 should be separated, the changes this pr makes are minimal and work, and can be easily reverted or extended for when someone pr's a better working solution.

**Comment by Becold** [comment]
What happened to this pull request ? Why is it not merged ?

**Comment by Chlodochar** [comment]
Each PR needs 2 contributors approving it before merging. I have approved it. Another contributor still needs to do so, that's why I added the `awaiting review` label.

**Comment by BaIthamel** [comment]
You need to deal with or at least respond to Chlo's inline comments before this will be merged.

File: static/js/map.js (hunk 1 above)
Comment (Studio-XP) [inline static/js/map.js]: Sorry, didn't notice it was fixed

File: static/js/map.js (hunk 2 above)
Comment (Studio-XP) [inline static/js/map.js]: Same here

File: static/js/sw.js (hunk 3 above)
Comment (Studio-XP) [inline static/js/sw.js]: The core of push notifications are serviceworkers. A sw.js file is required in order for them to work. A blank sw.js will introduce basic functionality, but we can add code later on to expand on the push notifications.

File: pogom/app.py (hunk 4 above)
Comment (Studio-XP) [inline pogom/app.py]: This line permits us to load the sw.js file from the document root (localhost:5000/sw.js). sw.js needs to be in the root of the site directory for it to work across the whole site.

Imagine you have define a cookie with a cookie path (directory). Those cookies will only work from that directory and deeper. Similar story here. A sw.js called from a subdirectory will only work from that directory and deeper. So if the sw.js file is located at localhost/example/, it will only work from the /example/ directory and on

**Comment by Studio-XP** [comment]
Sorry, thought I did in the comment under the SSL reference 594.

**Comment by nunopitbull** [comment]
Is this possible to copy paste on top of:
https://github.com/PokemonGoMap/PokemonGo-Map/pull/883

?

If not, how can I apply this without breaking the script above mentioned?

**Comment by Studio-XP** [comment]
@nunopitbull, I don't see any conflicts with the 2 sets of code. You should be able to modify it without any problems :)

**Comment by nunopitbull** [comment]
I mean, i'm a newbie. I just know how to grab on your RAW and copy/paste on the other one. Won't that delete the changes made by the "Lured pokestops" script?

**Comment by Studio-XP** [comment]
Yes, you cannot grab the raw version. Just follow the changes

**Comment by nunopitbull** [comment]
http://prntscr.com/c903of

**Comment by Studio-XP** [comment]
Make sure that the indentation on that line is the same as the one above it. Python is real particular on it

**Comment by nunopitbull** [comment]
Whats a identention? :P

**Comment by Studio-XP** [comment]
Yes. by indentation, i mean using the [ tab ] key. If you highlight the line above, you'll notice that it's not spaces, but instead tabs. Python doesn't use curly brackets `{ }` for statement blocks (if, else, function(), etc...), as javascript/c languages do, but instead relies on proper, consistent indentation. Since the python code on this project uses tabs for indentation, you must use tabs too. Just make sure that there are no spaces before my lines of code, and that you use the [ tab ] key instead. Do the same with the `def static_from_root()` function towards the bottom

**Comment by nunopitbull** [comment]
Same error, I used tab now

**Comment by Studio-XP** [comment]
He's running fine now, was using wrong line spacing in app.py

**Comment by khag7** [comment]
@Studio-XP have you looked at notification actions at all? It's pretty neat, I think it would go well with this feature (as another PR, sure, but along the same lines). See my PR which won't get merged: https://github.com/PokemonGoMap/PokemonGo-Map/pull/866

**Comment by Studio-XP** [comment]
Yes, it is still relevant. I may have to open a new pr due to the changes to map.js a few weeks ago though

**Comment by Studio-XP** [comment]
Somehow this pulled my gym marker pr into it.  I'm gonna open a new clean pr
